pkg/cloudprovider/provider/vsphere/provider.go

```diff
@@ -376,7 +408,19 @@ func (p *provider) Cleanup(machine *v1alpha1.Machine, data *cloudprovidertypes.P
 			}
 		}
 	}
-
+	var props mo.VirtualMachine
+	// Obtain VM properties
+	if err := virtualMachine.Properties(ctx, virtualMachine.Reference(), nil, &props); err != nil {
```

Comment:
Can you please explain why this is needed?

Reply:
This is to get the datastore where the VM is located, before we had the datastore in the config now we are not sure to have it (i.e. if the datastore cluster is specified instead) then I added this code to get the actual VM datastore location.

Reply:
As I said in another comment, you have already the value of the resource that you are trying reach and in order to make sure what you have is a datastore or a datastore cluster just check which one is set. This is a canonical workflow where the source of truth is coming from the configurations(especially when it comes to storage!)

Reply:
refer to the reply I thin in the other comment